astrogrism/simulate/simulate.py

```diff
@@ -0,0 +1,90 @@
+from os import environ
+from pathlib import Path
+from shutil import copy
+from tarfile import open as tar_open
+from tempfile import gettempdir
+from warnings import warn
+
+from astropy.utils.data import download_file
+from synphot import Observation
+
+
+def generate_simulation_spectrum(grism, detector=None, temp_path=gettempdir()):
+    """
+    Initializes and uses STSynphot to generate a Vega spectrum within the bandpass of a given grism
+
+    Parameters
+    ----------
+    grism : str
+        String representation of one of the four supported HST Grisms
+        Valid grisms: G141, G102, G280, G800L
+
+    detector : int
+        For detectors with multiple chips, specifies which chip to simulate
+        Only useful for G280 and G800L Grisms
+
+    temp_path : str
+        Path to download necessary files for STSynphot. Fallsback to Python's
+        default temporary folder location
+
+    """
+    SIM_DATA_DIR = Path(temp_path) / "astrogrism_simulation_files"
```

Comment:
Personally, I think that the default download path should be somewhere within the astrogrism package. The default python temp directory is just too annoying for a basic user to track down should they want to delete the files, IMO (I wanted to delete the files and it even took me a minute to track them down). I would guess, and could be totally wrong, that even with a pip download people are more used to looking in their python `site-packages` than in something like `/var/folders/ds/q5jd1xq53l9djq7lcc029d240002hz/`.

The default could be something like `pkg_resources.resource_filename("astrogrism", "simulate/")`, but maybe an even better answer is to have the path to download files in be a required argument?

Reply:
I'm also willing to be convinced that the current way using the tempdir is the best way 😁

Reply:
Here's my two cents about tempfiles: I think the user should NEVER be aware of these files; these are files that astrogrism needs to run, and are no different than javascript libraries for your webbrowser. At least on Windows, the files in the temp directory are intended to be immediate use and are subject to the OS' garbage collection at any point, so they shouldn't overstay their welcome. You don't really think twice about where Chrome downloads its files in the background, right?

Reply:
You can also use https://docs.astropy.org/en/latest/api/astropy.utils.data.get_pkg_data_path.html#astropy.utils.data.get_pkg_data_path if you want it somewhere in the package, but becareful not to accidentally check them into version control.

Reply:
Astronomers also like to do stuff on planes and on mountains with no internet, so as long as you also provide a "offline mode," where they can pre-download these files...

Reply:
> becareful not to accidentally check them into version control

I think the temp folder solves this problem by being entirely transparent to the user 🤷

Reply:
> provide a "offline mode," where they can pre-download these files...

Okay, that's a good point, and entirely against what I just said here: https://github.com/spacetelescope/astrogrism/pull/18#discussion_r824117590

Touche 😅

Reply:
@pllim That's a better example of what I was thinking - I think we want to err on the side of keeping these files around rather than needing to re-download them if the temp directory gets cleaned up. And using `pkg_resources` or that astropy utility means (I think) that they would get removed if astrogrism is uninstalled.

Reply:
If you use functions from `astropy.utils.data` and opt to cache, they get cached to where `astropy` stuff goes. It is documented at https://docs.astropy.org/en/latest/utils/data.html

Reply:
> err on the side of keeping these files
One subtlety that might be useful: the actual zip folder is cached using `astropy.utils.download_file` (as per Pey Lian's suggestion here: https://github.com/spacetelescope/astrogrism/pull/18#discussion_r816131099). This section is only for the unfolding of that archive. I'd argue the the files should only be unfolded when they're needed, then cleaned up when they're not needed. Having the cache from `download_file` removes the need to constantly redownload; just a matter of eating up the user's disk space

Reply:
BUT in the case if the user does want to keep them around, they can provide their own path now and point to it next time they run the simulation. But if they don't, then I'd advocate for being COMPLETELY invisible to the user

Reply:
Ahh, I see, I see. Clearly I didn't grok all the previous conversation...this is fine then. Explains why I deleted those files and then it took 7 seconds instead of 45 the second time.

Reply:
In the case where user already has stsynphot stuff set up, all this downloading is unnecessary? For example, if you are on any of the science server, all you have to do is point your `PYSYN_CDBS` to `/grp/redcat/trds/` (or something like that, they keep renaming it) without needing to download anything.

Reply:
The nice thing is I have the code check what files are missing from the archives and only unzip the ones that are necessary. So if the user provides their own pre-prepared path, astrogrism will just take it and write out PYSYN_CDBS to the environment variables for them, see that the files are already there, and move onto the simulation